File: src/ru/hudspeed/pro/Geo.java
package ru.gpsantiradar.app;

public final class Geo {
    private static final double EARTH_RADIUS_M = 6371000.0;

    private Geo() {}

    public static double distanceMeters(double lat1, double lon1, double lat2, double lon2) {
        double p1 = Math.toRadians(lat1);
        double p2 = Math.toRadians(lat2);
        double dp = Math.toRadians(lat2 - lat1);
        double dl = Math.toRadians(lon2 - lon1);
        double a = Math.sin(dp / 2) * Math.sin(dp / 2)
                + Math.cos(p1) * Math.cos(p2) * Math.sin(dl / 2) * Math.sin(dl / 2);
        return EARTH_RADIUS_M * 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
    }

    public static float bearing(double lat1, double lon1, double lat2, double lon2) {
        double p1 = Math.toRadians(lat1);
        double p2 = Math.toRadians(lat2);
        double dl = Math.toRadians(lon2 - lon1);
        double y = Math.sin(dl) * Math.cos(p2);
        double x = Math.cos(p1) * Math.sin(p2) - Math.sin(p1) * Math.cos(p2) * Math.cos(dl);
        return normalize((float) Math.toDegrees(Math.atan2(y, x)));
    }

    public static float angleDifference(float a, float b) {
        float d = Math.abs(normalize(a) - normalize(b));
        return d > 180f ? 360f - d : d;
    }

    public static float normalize(float angle) {
        float result = angle % 360f;
        return result < 0 ? result + 360f : result;
    }
}

File: src/main/kotlin/ru/gpsantiradar/app/Geo.kt
package ru.gpsantiradar.app

import kotlin.math.*

object Geo {
    private const val EARTH_RADIUS_M = 6371000.0

    fun distanceMeters(lat1: Double, lon1: Double, lat2: Double, lon2: Double): Double {
        val p1 = Math.toRadians(lat1)
        val p2 = Math.toRadians(lat2)
        val dp = Math.toRadians(lat2 - lat1)
        val dl = Math.toRadians(lon2 - lon1)
        val a = sin(dp / 2) * sin(dp / 2) + cos(p1) * cos(p2) * sin(dl / 2) * sin(dl / 2)
        return EARTH_RADIUS_M * 2 * atan2(sqrt(a), sqrt(1 - a))
    }

    fun bearing(lat1: Double, lon1: Double, lat2: Double, lon2: Double): Float {
        val p1 = Math.toRadians(lat1)
        val p2 = Math.toRadians(lat2)
        val dl = Math.toRadians(lon2 - lon1)
        val y = sin(dl) * cos(p2)
        val x = cos(p1) * sin(p2) - sin(p1) * cos(p2) * cos(dl)
        return normalize(Math.toDegrees(atan2(y, x)).toFloat())
    }

    fun angleDifference(a: Float, b: Float): Float {
        val difference = abs(normalize(a) - normalize(b))
        return if (difference > 180f) 360f - difference else difference
    }

    fun angleDifferenceSigned(first: Float, second: Float): Float {
        var difference = (second - first) % 360f
        if (difference > 180f) difference -= 360f
        if (difference < -180f) difference += 360f
        return difference
    }

    fun normalize(angle: Float): Float {
        val result = angle % 360f
        return if (result < 0) result + 360f else result
    }
}
